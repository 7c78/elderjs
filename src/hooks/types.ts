export type HookOptions = {
  hook: string;
  name: string;
  description: string;
  priority: Number;
  run: Run;
  $$meta?: {
    type: string;
    addedBy: string;
  };
};

interface Run {
  (input: any): any | Promise<any>;
}
