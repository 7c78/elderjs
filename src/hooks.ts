import path from 'path';
import fs from 'fs-extra';
import { parseBuildPerf } from './utils';
import externalHelpers from './externalHelpers';
import { HookOptions } from './hookInterface/types';

const hooks: Array<HookOptions> = [
  {
    hook: 'bootstrap',
    name: 'elderAddExternalHelpers',
    description: 'Adds external helpers to helpers object',
    priority: 100,
    run: async ({ helpers, query, settings }) => {
      let additionalHelpers = {};
      try {
        additionalHelpers = await externalHelpers({ helpers, query, settings });
      } catch (err) {
        console.error(err);
      }
      if (additionalHelpers) {
        return {
          helpers: {
            ...helpers,
            ...additionalHelpers,
          },
        };
      }
      return null;
    },
  },
  {
    hook: 'stacks',
    name: 'elderAddMetaCharsetToHead',
    description: `Adds <meta charset="UTF-8" /> to the head.`,
    priority: 1,
    run: async ({ headStack }) => {
      return {
        headStack: [
          ...headStack,
          {
            source: 'elderAddMetaCharsetToHead',
            string: `<meta charset="UTF-8" />`,
            priority: 1,
          },
        ],
      };
    },
  },
  {
    hook: 'stacks',
    name: 'elderAddMetaViewportToHead',
    description: `Adds <meta name="viewport" content="width=device-width, initial-scale=1" /> to the head.`,
    priority: 10,
    run: async ({ headStack }) => {
      return {
        headStack: [
          ...headStack,
          {
            source: 'elderAddMetaViewportToHead',
            string: `<meta name="viewport" content="width=device-width, initial-scale=1" />`,
            priority: 10,
          },
        ],
      };
    },
  },
  {
    hook: 'stacks',
    name: 'elderAddDefaultIntersectionObserver',
    description: 'Sets up the default polyfill for the intersection observer',
    priority: 1,
    run: async ({ beforeHydrateStack, settings }) => {
      if (settings && settings.locations && {}.hasOwnProperty.call(settings.locations, 'intersectionObserverPoly')) {
        if (settings.locations.intersectionObserverPoly) {
          return {
            beforeHydrateStack: [
              {
                source: 'elderAddDefaultIntersectionObserver',
                string: `<script type="text/javascript">
      if (!('IntersectionObserver' in window)) {
          var script = document.createElement("script");
          script.src = "${settings.locations.intersectionObserverPoly}";
          document.getElementsByTagName('head')[0].appendChild(script);
      };
      </script>`,
                priority: 1,
              },
              ...beforeHydrateStack,
            ],
          };
        }
      } else {
        console.log(
          'Not injecting intersection observer polyfill. To not see this warning set locations.intersectionObserverPoly = false in elder.config.js.',
        );
      }
      return null;
    },
  },
  {
    hook: 'stacks',
    name: 'elderAddSystemJs',
    description: 'AddsSystemJs to beforeHydrateStack also add preloading of systemjs to the headStack.',
    priority: 1,
    run: async ({ beforeHydrateStack, headStack, settings }) => {
      if (settings && settings.locations && {}.hasOwnProperty.call(settings.locations, 'systemJs')) {
        if (settings.locations.systemJs) {
          return {
            beforeHydrateStack: [
              {
                source: 'elderAddSystemJs',
                string: `<script data-name="systemjs" src="${settings.locations.systemJs}"></script>`,
                priority: 2,
              },
              ...beforeHydrateStack,
            ],

            headStack: [
              {
                source: 'elderAddSystemJs',
                string: `<link rel="preload" href="${settings.locations.systemJs}" as="script">`,
                priority: 2,
              },
              ...headStack,
            ],
          };
        }
      } else {
        console.log(
          'Not injecting systemjs. To not see this warning set locations.systemJs = false in elder.config.js.',
        );
      }
      return null;
    },
  },
  {
    hook: 'compileHtml',
    name: 'elderCompileHtml',
    description: 'Creates an HTML string out of the Svelte layout and stacks.',
    priority: 50,
    run: async ({ request, headString, footerString, layoutHtml, htmlString }) => {
      return {
        htmlString: `<!DOCTYPE html><html lang="en"><head>${headString}</head><body class="${request.route}">${layoutHtml}${footerString}</body></html>`,
      };
    },
  },

  {
    hook: 'error',
    name: 'elderConsoleLogErrors',
    description: 'Log any errors to the console.',
    priority: 100,
    run: async ({ errors }) => {
      console.error(errors);
    },
  },
  {
    hook: 'requestComplete',
    name: 'elderWriteHtmlFileToPublic',
    description: 'Write the html output to public.',
    priority: 100,
    run: async ({ settings, request, htmlString, errors }) => {
      if (settings.build) {
        const file = path.resolve(process.cwd(), `${settings.locations.public}${request.permalink}/index.html`);
        try {
          fs.outputFileSync(file, htmlString);
        } catch (e) {
          console.log(e);
          return {
            errors: [...errors, e],
          };
        }
      }
      return null;
    },
  },
  {
    hook: 'requestComplete',
    name: 'elderDisplayRequestTime',
    description: 'Page generating timings and logging.',
    priority: 50,
    run: async ({ timings, request, settings }) => {
      if (!settings.build && process.env.NODE_ENV !== 'production') {
        console.log(`${Math.round(timings.slice(-1)[0].duration * 10) / 10}ms: \t ${request.permalink}`);
        if (settings.debug.performance) {
          const outTimings = [...timings];
          const display = outTimings.sort((a, b) => a.duration - b.duration).map((t) => ({ ...t, ms: t.duration }));

          console.table(display, ['name', 'ms']);
        }
      }
    },
  },
  {
    hook: 'buildComplete',
    name: 'elderShowParsedBuildTimes',
    description: 'A breakdown of the average times of different stages of the build.',
    priority: 50,
    run: async ({ timings, settings }) => {
      if (settings.debug.performance) {
        const buildTimings = parseBuildPerf(timings);
        console.log(buildTimings);
      }
    },
  },
  {
    hook: 'buildComplete',
    name: 'elderWriteBuildErrors',
    description: 'Writes out any errors of a build to a JSON file in the ___ELDER___ folder.',
    priority: 50,
    run: async ({ errors, settings }) => {
      if (errors && errors.length > 0) {
        const buildOutputLocation = path.resolve(process.cwd(), `./___ELDER___/build-${Date.now()}.json`);
        console.log(`Writing details on the ${errors.length} build errors to: ${buildOutputLocation}`);
        fs.writeJSONSync(buildOutputLocation, { errors, settings });
      }
    },
  },
];

export default hooks;
