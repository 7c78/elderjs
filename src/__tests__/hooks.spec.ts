import hooks from '../hooks';

jest.mock('../externalHelpers', () => () => Promise.resolve({ permalink: jest.fn() }));

process.cwd = () => 'test';

jest.mock('path', () => ({
  resolve: (...strings) => strings.join('/').replace('./', '').replace('//', '/'),
  posix: () => ({ dirname: () => '' }),
}));

jest.mock('fs-extra', () => ({
  writeJSONSync: jest.fn(),
  outputFileSync: jest
    .fn()
    .mockImplementationOnce(() => {})
    .mockImplementationOnce(() => {
      throw new Error('Failed to write');
    }),
}));

describe('#hooks', () => {
  it('has valid priority', () => {
    expect(hooks.filter((h) => h.priority < 1 && h.priority > 100)).toEqual([]);
  });
  it('matchesSnapshot', () => {
    expect(hooks).toMatchSnapshot();
  });
  it('elderAddExternalHelpers', async () => {
    const hook = hooks.find((h) => h.name === 'elderAddExternalHelpers');
    expect(await hook.run({ helpers: { old: jest.fn() }, query: {}, settings: {} })).toMatchSnapshot();
  });
  it('elderProcessShortcodes', async () => {
    const hook = hooks.find((h) => h.name === 'elderAddExternalHelpers');
    expect(await hook.run({ helpers: { old: jest.fn() }, query: {}, settings: {} })).toMatchSnapshot();

    // TODO!!!
  });
  it('elderAddMetaCharsetToHead', async () => {
    const hook = hooks.find((h) => h.name === 'elderAddMetaCharsetToHead');
    expect(await hook.run({ headStack: [] })).toMatchSnapshot();
  });
  it('elderAddMetaViewportToHead', async () => {
    const hook = hooks.find((h) => h.name === 'elderAddMetaViewportToHead');
    expect(await hook.run({ headStack: [] })).toMatchSnapshot();
  });
  it('elderAddDefaultIntersectionObserver', async () => {
    const hook = hooks.find((h) => h.name === 'elderAddDefaultIntersectionObserver');
    expect(await hook.run({ beforeHydrateStack: [] })).toMatchSnapshot();
  });
  it('elderAddSystemJs', async () => {
    const hook = hooks.find((h) => h.name === 'elderAddSystemJs');
    expect(await hook.run({ beforeHydrateStack: [], headStack: [] })).toMatchSnapshot();
  });

  it('elderCompileHtml', async () => {
    const hook = hooks.find((h) => h.name === 'elderCompileHtml');
    expect(
      await hook.run({
        request: { route: 'test' },
        headString: 'head',
        footerString: 'footer',
        layoutHtml: 'layout',
      }),
    ).toStrictEqual({
      htmlString: '<!DOCTYPE html><html lang="en"><head>head</head><body class="test">layoutfooter</body></html>',
    });
  });

  it('elderConsoleLogErrors', async () => {
    const hook = hooks.find((h) => h.name === 'elderConsoleLogErrors');
    expect(await hook.run({ request: { permalink: '/foo' }, errors: ['foo', 'bar'] })).toBe(undefined);
  });
  it('elderWriteHtmlFileToPublic', async () => {
    const hook = hooks.find((h) => h.name === 'elderWriteHtmlFileToPublic');
    expect(
      await hook.run({
        request: { permalink: '/foo' },
        htmlString: '<html>string</html>',
        errors: [],
        settings: {},
      }),
    ).toBe(null);
    expect(
      await hook.run({
        request: { permalink: '/foo' },
        htmlString: '<html>string</html>',
        errors: [],
        settings: { build: './build', locations: { public: './public' } },
      }),
    ).toBe(null);
    expect(
      await hook.run({
        request: { permalink: '/foo' },
        htmlString: '<html>string</html>',
        errors: [],
        settings: { build: './build', locations: { public: './public' } },
      }),
    ).toEqual({ errors: [new Error('Failed to write')] });
  });
  it('elderDisplayRequestTime', async () => {
    const hook = hooks.find((h) => h.name === 'elderDisplayRequestTime');
    expect(
      await hook.run({
        request: { permalink: '/foo' },
        timings: [
          { name: 'foo', duration: 500 },
          { name: 'bar', duration: 250 },
        ],
        settings: { debug: { performance: true } },
      }),
    ).toBe(undefined);
  });
  it('elderShowParsedBuildTimes', async () => {
    const hook = hooks.find((h) => h.name === 'elderShowParsedBuildTimes');
    expect(
      await hook.run({
        timings: [
          [
            { name: 'foo', duration: 500 },
            { name: 'bar', duration: 250 },
          ],
          [
            { name: 'foo', duration: 2500 },
            { name: 'bar', duration: 0 },
          ],
        ],
        settings: { debug: { performance: true } },
      }),
    ).toBe(undefined);
  });
  it('elderWriteBuildErrors', async () => {
    const hook = hooks.find((h) => h.name === 'elderWriteBuildErrors');
    expect(
      await hook.run({
        errors: ['error1', 'error2'],
        settings: { debug: { performance: true } },
      }),
    ).toBe(undefined);
  });
});
