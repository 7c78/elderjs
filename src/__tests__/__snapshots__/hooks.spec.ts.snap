// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#hooks elderAddDefaultIntersectionObserver 1`] = `
Object {
  "beforeHydrateStack": Array [
    Object {
      "priority": 1,
      "source": "elderAddDefaultIntersectionObserver",
      "string": "<script type=\\"text/javascript\\">
      if (!('IntersectionObserver' in window)) {
          var script = document.createElement(\\"script\\");
          script.src = \\"foo\\";
          document.getElementsByTagName('head')[0].appendChild(script);
      };
      </script>",
    },
  ],
}
`;

exports[`#hooks elderAddExternalHelpers 1`] = `
Object {
  "helpers": Object {
    "old": [MockFunction],
    "permalink": [MockFunction],
  },
}
`;

exports[`#hooks elderAddMetaCharsetToHead 1`] = `
Object {
  "headStack": Array [
    Object {
      "priority": 1,
      "source": "elderAddMetaCharsetToHead",
      "string": "<meta charset=\\"UTF-8\\" />",
    },
  ],
}
`;

exports[`#hooks elderAddMetaViewportToHead 1`] = `
Object {
  "headStack": Array [
    Object {
      "priority": 10,
      "source": "elderAddMetaViewportToHead",
      "string": "<meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\" />",
    },
  ],
}
`;

exports[`#hooks elderAddSystemJs 1`] = `
Object {
  "beforeHydrateStack": Array [
    Object {
      "priority": 2,
      "source": "elderAddSystemJs",
      "string": "<script data-name=\\"systemjs\\" src=\\"foo\\"></script>",
    },
  ],
  "headStack": Array [
    Object {
      "priority": 2,
      "source": "elderAddSystemJs",
      "string": "<link rel=\\"preload\\" href=\\"foo\\" as=\\"script\\">",
    },
  ],
}
`;

exports[`#hooks matchesSnapshot 1`] = `
Array [
  Object {
    "description": "Adds external helpers to helpers object",
    "hook": "bootstrap",
    "name": "elderAddExternalHelpers",
    "priority": 100,
    "run": [Function],
  },
  Object {
    "description": "Adds <meta charset=\\"UTF-8\\" /> to the head.",
    "hook": "stacks",
    "name": "elderAddMetaCharsetToHead",
    "priority": 1,
    "run": [Function],
  },
  Object {
    "description": "Adds <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\" /> to the head.",
    "hook": "stacks",
    "name": "elderAddMetaViewportToHead",
    "priority": 10,
    "run": [Function],
  },
  Object {
    "description": "Sets up the default polyfill for the intersection observer",
    "hook": "stacks",
    "name": "elderAddDefaultIntersectionObserver",
    "priority": 1,
    "run": [Function],
  },
  Object {
    "description": "AddsSystemJs to beforeHydrateStack also add preloading of systemjs to the headStack.",
    "hook": "stacks",
    "name": "elderAddSystemJs",
    "priority": 1,
    "run": [Function],
  },
  Object {
    "description": "Creates an HTML string out of the Svelte layout and stacks.",
    "hook": "compileHtml",
    "name": "elderCompileHtml",
    "priority": 100,
    "run": [Function],
  },
  Object {
    "description": "Log any errors to the console.",
    "hook": "error",
    "name": "elderConsoleLogErrors",
    "priority": 100,
    "run": [Function],
  },
  Object {
    "description": "Write the html output to public.",
    "hook": "requestComplete",
    "name": "elderWriteHtmlFileToPublic",
    "priority": 100,
    "run": [Function],
  },
  Object {
    "description": "Page generating timings and logging.",
    "hook": "requestComplete",
    "name": "elderDisplayRequestTime",
    "priority": 50,
    "run": [Function],
  },
  Object {
    "description": "A breakdown of the average times of different stages of the build.",
    "hook": "buildComplete",
    "name": "elderShowParsedBuildTimes",
    "priority": 50,
    "run": [Function],
  },
  Object {
    "description": "Writes out any errors of a build to a JSON file in the ___ELDER___ folder.",
    "hook": "buildComplete",
    "name": "elderWriteBuildErrors",
    "priority": 50,
    "run": [Function],
  },
]
`;
